sandbox/apparmor/notify/listener/listener.go

```diff
@@ -0,0 +1,251 @@
+	if err != nil {
+		if os.IsNotExist(err) {
+			return nil, ErrListenerNotSupported
+		}
+		return nil, fmt.Errorf("cannot open %q: %v", path, err)
+	}
+
+	msg := notify.MsgNotificationFilter{ModeSet: notify.APPARMOR_MODESET_USER}
+	data, err := msg.MarshalBinary()
+	if err != nil {
+		notifyFile.Close()
+		return nil, err
+	}
+	ioctlBuf := notify.BytesToIoctlRequestBuffer(data)
+	_, err = notifyIoctl(notifyFile.Fd(), notify.APPARMOR_NOTIF_SET_FILTER, ioctlBuf)
+	if err != nil {
+		notifyFile.Close()
+		return nil, fmt.Errorf("cannot notify ioctl to modeset user on %q: %v", path, err)
+	}
+
+	poll, err := epoll.Open()
+	if err != nil {
+		notifyFile.Close()
+		return nil, fmt.Errorf("cannot open epoll file descriptor: %v", err)
+	}
+	// XXX: Do we need a notification for Writable, to send responses back?
+	if err := poll.Register(int(notifyFile.Fd()), epoll.Readable); err != nil {
+		notifyFile.Close()
+		poll.Close()
+		return nil, fmt.Errorf("cannot register epoll on %q: %v", path, err)
+	}
+
+	listener := &Listener{
+		R: make(chan *Request),
+		E: make(chan error),
+
+		notifyFile: notifyFile,
+		poll:       poll,
+	}
+	return listener, nil
+}
+
+func (l *Listener) decodeAndDispatchRequest(buf []byte, tomb *tomb.Tomb) error {
+	var nmsg notify.MsgNotification
+	if err := nmsg.UnmarshalBinary(buf); err != nil {
+		return err
+	}
+	// What kind of notification message did we get?
+	switch nmsg.NotificationType {
+	case notify.APPARMOR_NOTIF_OP:
+		var omsg notify.MsgNotificationOp
+		if err := omsg.UnmarshalBinary(buf); err != nil {
+			return err
+		}
+		// What kind of operation notification did we get?
+		switch omsg.Class {
+		case notify.AA_CLASS_FILE:
+			var fmsg notify.MsgNotificationFile
+			if err := fmsg.UnmarshalBinary(buf); err != nil {
+				return err
+			}
+			// log.Printf("notification request: %#v\n", fmsg)
+			req := newRequest(l, &fmsg)
+			l.R <- req
+			tomb.Go(func() error {
+				l.waitAndRespond(req, &fmsg)
+				return nil
+			})
+		default:
+			return fmt.Errorf("unsupported mediation class: %v", omsg.Class)
+		}
+	default:
+		return fmt.Errorf("unsupported notification type: %v", nmsg.NotificationType)
+	}
+	return nil
+}
+
+func (l *Listener) waitAndRespond(req *Request, msg *notify.MsgNotificationFile) {
+	resp := notify.ResponseForRequest(&msg.MsgNotification)
+	resp.MsgNotification.Error = 0 // ignored in responses
+	resp.MsgNotification.NoCache = 1
+	if allow := <-req.YesNo; allow {
+		resp.Allow = msg.Allow | msg.Deny
+		resp.Deny = 0
+		resp.Error = 0
+	} else {
+		resp.Allow = 0
+		resp.Deny = msg.Deny
+		resp.Error = msg.Error
+		// msg.Error is field from MsgNotificationResponse, and is unused.
+		// msg.MsgNotification.Error is also ignored in responses.
+	}
+	//log.Printf("notification response: %#v\n", resp)
+	if err := l.encodeAndSendResponse(&resp); err != nil {
+		l.fail(err)
+	}
+}
+
+func (l *Listener) encodeAndSendResponse(resp *notify.MsgNotificationResponse) error {
```

Comment:
(nitpick/questionable advise) given that this helper is used only once (so far?) and that it is very small and obvious I wonder if it sould just be folded nto waitAndRespond:
```go
@@ -180,19 +184,14 @@ func (l *Listener) waitAndRespond(req *Request, msg *notify.MsgNotificationFile)
                // msg.MsgNotification.Error is also ignored in responses.
        }
        //log.Printf("notification response: %#v\n", resp)
-       if err := l.encodeAndSendResponse(&resp); err != nil {
-               l.fail(err)
-       }
-}
-
-func (l *Listener) encodeAndSendResponse(resp *notify.MsgNotificationResponse) error {
        buf, err := resp.MarshalBinary()
        if err != nil {
-               return err
+               l.fail(err)
        }
        ioctlBuf := notify.BytesToIoctlRequestBuffer(buf)
-       _, err = notifyIoctl(l.notifyFile.Fd(), notify.APPARMOR_NOTIF_SEND, ioctlBuf)
-       return err
+       if _, err = notifyIoctl(l.notifyFile.Fd(), notify.APPARMOR_NOTIF_SEND, ioctlBuf); err != nil {
+               l.fail(err)
+       }
 }
 
 func (l *Listener) runOnce(tomb *tomb.Tomb) error {
```

Reply:
`waitAndRespond` takes a `MsgNotificationFile` as a parameter, and setting Allow/Deny is dependent on it being a file, but `encodeAndSendResponse` just takes a ready-to-send response and could be used by any future `waitAndRespond` function for other message types, so I think it's reasonable to keep the helper function as-is. Really, what I'd rather combine is `handleRequestAaClassFile` and `waitAndRespond`, as both are for `MsgNotificationFile` types. But: it's also important that `waitAndRespond` occurs in its own tomb-managed goroutine, and we don't want to mess with spawning goroutines in the switch statement of class types. So, I think the flow of `decodeAndDispatchRequest` -> `handleRequestAaClass<class>` -> `tomb.Go(waitAndRespondAaClass<class>)` -> (common) `encodeAndSendResponse` makes sense.